### PR TITLE
CASMTRIAGE-7594 - improve node acquisition and rebalance.

### PR DESCRIPTION
**Pull request by dlaine-hpe**
## Summary and Scope

There was a problem where a worker node's console was not being monitored by anyone because it was getting assigned/unassigned to the console-node pod that was running on that worker and the other pod didn't have capacity to pick it up. The end result was it was getting picked up and dropped continually while forcing the conmand process to keep getting killed (resulting in incomplete logging of all the nodes in that pod).

The fix was to rework how node acquisition and rebalancing happens to be much more stable and aware of the current status of the services. This fix requires changes in all three console repos.

At the same time I added a change to watch the permissions on the log files. There were problems where something was setting the files to unwritable and losing logging.

## Issues and Related PRs
* Resolves [CASMTRIAGE-7594](https://jira-pro.it.hpe.com:8443/browse/CASMTRIAGE-7594)
* Resolves [CASMCMS-9126](https://jira-pro.it.hpe.com:8443/browse/CASMCMS-9126)

### Tested on:
  * `Surtur`

### Test description:

I installed all 3 new versions of the console services and monitored the node acquisition through pod rollout, then forced pod failures to insure the remaining pods picked up the orphaned nodes, then correctly rebalanced when all console-node pods were back up and running.

- Were the install/upgrade-based validation checks/tests run (goss tests/install-validation doc)? N
- Were continuous integration tests run? If not, why? N
- Was upgrade tested? If not, why? N
- Was downgrade tested? If not, why? N
- Were new tests (or test issues/Jiras) created for this change? N

## Risks and Mitigations

This is a medium size change, but I spent several days testing in all situations I could think up.

## Pull Request Checklist
- [X] Version number(s) incremented, if applicable
- [X] Copyrights updated
- [X] License file intact
- [X] Target branch correct
- [X] CHANGELOG.md updated
- [X] Testing is appropriate and complete, if applicable

### CODE REVIEW

File: src/console_node/nodes.go [hunk 1]
```diff
@@ -107,54 +129,151 @@ func pinNumNodes(numAsk, numMax int) int {
...
 		}
-	} else {
-		log.Printf("Skipping acquire - at capacity. CurRvr:%d, TarRvr:%d, CurMtn:%d, TarMtn:%d",
-			len(currentRvrNodes), targetRvrNodes, len(currentMtnNodes)+len(currentPdsNodes), targetMtnNodes)
+		log.Printf(" Added River:%d, Mountain:%d, Paradise:%d", newRvr, newMtn, newPds)
 	}
 
 	// See if we have too many nodes
-	if rebalanceNodes() {
+	if rebalanceNodes(deltaRvr, deltaMtn) {
```
Comment (jsollom-hpe) [inline src/console_node/nodes.go:276]: So, we have mountain, river, and paradise nodes, but you're only dealing with deltaRvr and deltaMtn. Should there be a deltaParadise or is that a subset of Mountain?

**Review by jsollom-hpe — APPROVED** [review]

File: src/console_node/nodes.go [hunk 2]
```diff
@@ -94,6 +95,27 @@ var newNodeLookupSec int = 30
...
+
+	// gather the names of all the current nodes being watched
+	var retVal []string
+	for key := range currentMtnNodes {
+		retVal = append(retVal, key)
+	}
+	for key := range currentRvrNodes {
+		retVal = append(retVal, key)
+	}
+	for key := range currentPdsNodes {
```
[the remark this answers is not in the document]
Comment (dlaine-hpe) [reply src/console_node/nodes.go:112]: "Paradise"

File: src/console_node/nodes.go [hunk 3]
```diff
@@ -107,54 +129,151 @@ func pinNumNodes(numAsk, numMax int) int {
...
+		}
+		if float64(currNumRvr) > (1.1 * float64(idealNumRvr)) {
+			log.Printf("  ** (float64(currNumRvr) > (1.1 * float64(idealNumRvr))) = true")
+			log.Printf("    float64(currNumRvr)=%f, (1.1 * float64(idealNumRvr))=%f", float64(currNumRvr), 1.1*float64(idealNumRvr))
+		}
+		// if we are under the ideal number, try to match
+		// if we are more than 10% over the ideal, give some back
+		// NOTE: there is a gap where we don't change the number to try and avoid bouncing
+		if (currNumRvr < idealNumRvr) || (float64(currNumRvr) > (1.1 * float64(idealNumRvr))) {
+			// get more to bring us up to the ideal number of nodes
```
[the remark this answers is not in the document]
Comment (dlaine-hpe) [reply src/console_node/nodes.go:202]: You are correct - comment was left over from before I combined the if statements.

File: src/console_node/nodes.go [hunk 4]
```diff
@@ -107,54 +129,151 @@ func pinNumNodes(numAsk, numMax int) int {
...
+			deltaRvr = idealNumRvr - currNumRvr
+			log.Printf("  ** Setting deltaRvr=%d", deltaRvr)
+		}
+		if (currNumMtn < idealNumMtn) || (float64(currNumMtn) > (1.1 * float64(idealNumMtn))) {
+			// get more to bring us up to the ideal number of nodes
```
[the remark this answers is not in the document]
Comment (dlaine-hpe) [reply src/console_node/nodes.go:207]: You are correct - comment was left over from before I combined the if statements.

File: src/console_node/nodes.go (hunk 1 above)
Comment (dlaine-hpe) [reply src/console_node/nodes.go:276]: Paradise is a subset of mountain. That is documented in a bunch of other places so didn't here.
At some point we will need to make the buckets more generic: 'ipmi' (river), 'ssh w/ key' (mountain), 'ssh w/ creds' (paradise), but that is for another day. Mountain and Paradise are both ssh based so need 'expect' and 'ssh' processes and use the same amount of resources.

**Pull request merged**